File: app/src/main/java/com/nfcalarmclock/NacCheckboxPreference.java
package com.nfcalarmclock;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.util.AttributeSet;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

/**
 * Preference that indicates repeating the alarm.
 */
public class NacCheckboxPreference
	extends Preference
	implements Preference.OnPreferenceClickListener,
		CompoundButton.OnCheckedChangeListener
{

	/**
	 * Check box button.
	 */
	private CheckBox mCheckBox;

	/**
	 * Check value.
	 */
	protected boolean mValue;

	/**
	 * Summary text when checkbox is enabled.
	 */
	protected String mSummaryOn;

	/**
	 * Summary text when checkbox is disabled.
	 */
	protected String mSummaryOff;

	/**
	 * Color state list of the checkbox.
	 */
	protected ColorStateList mColorStateList;

	/**
	 * Binding view holder flag.
	 */
	protected boolean mBindFlag;

	/**
	 * Default value.
	 */
	protected static final boolean DEFAULT_VALUE = true;

	/**
	 */
	public NacCheckboxPreference(Context context)
	{
		super(context);
		this.init(context, null);
	}

	/**
	 */
	public NacCheckboxPreference(Context context, AttributeSet attrs)
	{
		super(context, attrs);
		this.init(context, attrs);
	}

	/**
	 */
	public NacCheckboxPreference(Context context, AttributeSet attrs, int style)
	{
		super(context, attrs, style);
		this.init(context, attrs);
	}

	/**
	 * @return The checked status.
	 */
	public boolean getChecked()
	{
		return this.mValue;
	}

	/**
	 */
	@Override
	public CharSequence getSummary()
	{
		return (this.mValue) ? this.mSummaryOn : this.mSummaryOff;
	}

	/**
	 * Initialize the preference.
	 */
	private void init(Context context, AttributeSet attrs)
	{
		setLayoutResource(R.layout.nac_preference_checkbox);
		setOnPreferenceClickListener(this);

		this.mColorStateList = null;
		this.mBindFlag = false;
		int[] array = new int[] { android.R.attr.summaryOn,
			android.R.attr.summaryOff };
		TypedArray ta = context.obtainStyledAttributes(attrs, array);

		try
		{
			this.mSummaryOn = ta.getString(0);
			this.mSummaryOff = ta.getString(1);
		}
		finally
		{
			ta.recycle();
		}
	}

	/**
	 */
	@Override
	public void onAttached()
	{
		super.onAttached();
		Context context = getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);
		int[][] states = new int[][] {
			new int[] {  android.R.attr.state_checked },
			new int[] { -android.R.attr.state_checked } };
		int[] colors = new int[] { shared.getThemeColor(), Color.LTGRAY };
		this.mColorStateList = new ColorStateList(states, colors);
	}

	/**
	 */
	@Override
	public void onBindViewHolder(PreferenceViewHolder holder)
	{
		super.onBindViewHolder(holder);

		this.mCheckBox = (CheckBox) holder.findViewById(R.id.widget);

		this.mCheckBox.setOnCheckedChangeListener(null);
		this.mCheckBox.setChecked(this.mValue);
		this.mCheckBox.setOnCheckedChangeListener(this);
		this.mCheckBox.setButtonTintList(this.mColorStateList);
	}

	/**
	 * Handle checkbox changes.
	 */
	@Override
	public void onCheckedChanged(CompoundButton button, boolean state)
	{
		if (!callChangeListener(state))
		{
			return;
		}

		this.mValue = state;

		this.mCheckBox.setChecked(state);
		setSummary(this.getSummary());
		persistBoolean(state);
		notifyDependencyChange(!state);
	}

	/**
	 * @return The default value.
	 */
	@Override
	protected Object onGetDefaultValue(TypedArray a, int index)
	{
		return a.getBoolean(index, DEFAULT_VALUE);
	}

	/**
	 * Allow users to select the whole preference to change the checkbox.
	 */
	@Override
	public boolean onPreferenceClick(Preference pref)
	{
		this.mCheckBox.performClick();
		return true;
	}

	/**
	 * Set the initial preference value.
	 */
	@Override
	protected void onSetInitialValue(Object defaultValue)
	{
		if (defaultValue == null)
		{
			this.mValue = getPersistedBoolean(this.mValue);
		}
		else
		{
			this.mValue = (boolean) defaultValue;

			persistBoolean(this.mValue);
		}
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/settings/preference/NacCheckboxPreference.kt
package com.nfcalarmclock.settings.preference

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.content.res.TypedArray
import android.graphics.Color
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences


/**
 * Preference that indicates repeating the alarm.
 */
@SuppressLint("ResourceType")
class NacCheckboxPreference(

	/**
	 * Context.
	 */
	context: Context,

	/**
	 * Attribute set.
	 */
	attrs: AttributeSet?

	// Constructor
) : Preference(context, attrs),

	// Interfaces
	Preference.OnPreferenceClickListener,
	CompoundButton.OnCheckedChangeListener
{

	/**
	 * The parent layout.
	 */
	private lateinit var parent: ViewGroup

	/**
	 * Check box button.
	 */
	private lateinit var checkBox: CheckBox

	/**
	 * Check value.
	 */
	var isChecked = false

	/**
	 * Summary text when checkbox is enabled.
	 */
	private var enabledSummary: String? = null

	/**
	 * Summary text when checkbox is disabled.
	 */
	private var disabledSummary: String? = null

	/**
	 * The enabled status that was buffered until the view is bound.
	 *
	 * Note: If it is null, the enabled status is not buffered.
	 */
	private var bufferedEnableStatus: Boolean? = null

	/**
	 * Constructor.
	 */
	init
	{
		// Set the layout
		layoutResource = R.layout.nac_preference_checkbox

		// Set the listener
		onPreferenceClickListener = this

		// Find the summary from the attributes set
		val array = intArrayOf(android.R.attr.summaryOn, android.R.attr.summaryOff)
		val ta = context.obtainStyledAttributes(attrs, array)

		// Set the summaries from the attributes
		try
		{
			enabledSummary = ta.getString(0)
			disabledSummary = ta.getString(1)
		}
		// Recycle the typed array
		finally
		{
			ta.recycle()
		}
	}

	/**
	 * Get the summary text.
	 *
	 * @return The summary text.
	 */
	override fun getSummary(): CharSequence?
	{
		// Enabled summary
		return if (isChecked)
			{
				enabledSummary
			}
			// Disabled summary
			else
			{
				disabledSummary
			}
	}

	/**
	 * Called when the view holder is bound.
	 */
	override fun onBindViewHolder(holder: PreferenceViewHolder)
	{
		// Super
		super.onBindViewHolder(holder)

		// Set the checkbox view
		parent = holder.itemView as ViewGroup
		checkBox = holder.findViewById(R.id.widget) as CheckBox

		// Set the checked status and sandwich it by unsetting and resetting
		// the listener so that it does not go off when the status is set
		checkBox.setOnCheckedChangeListener(null)
		checkBox.isChecked = isChecked
		checkBox.setOnCheckedChangeListener(this)

		// Setup color
		setupCheckBoxColor()

		// Check if the enabled status was buffered
		if (bufferedEnableStatus != null)
		{
			this.isEnabled = bufferedEnableStatus!!
			bufferedEnableStatus = null
		}
	}

	/**
	 * Handle checkbox changes.
	 */
	override fun onCheckedChanged(button: CompoundButton, state: Boolean)
	{
		// Call the listener and check if the call was canceled
		if (!callChangeListener(state))
		{
			return
		}

		// Set the checked status
		isChecked = state
		checkBox.isChecked = state

		// Set the summary
		summary = this.summary

		// Persist the state
		persistBoolean(state)

		// Don't know what this does
		notifyDependencyChange(!state)
	}

	/**
	 * Get the default value.
	 *
	 * @return The default value.
	 */
	override fun onGetDefaultValue(a: TypedArray, index: Int): Any
	{
		return a.getBoolean(index, DEFAULT_VALUE)
	}

	/**
	 * Allow users to select the whole preference to change the checkbox.
	 */
	override fun onPreferenceClick(pref: Preference): Boolean
	{
		// Perform the click
		checkBox.performClick()

		return true
	}

	/**
	 * Set the initial preference value.
	 */
	override fun onSetInitialValue(defaultValue: Any?)
	{
		// Check if the default value is null
		if (defaultValue == null)
		{
			isChecked = getPersistedBoolean(isChecked)
		}
		// Convert the default value
		else
		{
			isChecked = defaultValue as Boolean

			persistBoolean(isChecked)
		}
	}

	/**
	 * Set whether the preference is enabled or not.
	 */
	override fun setEnabled(enabled: Boolean)
	{
		// Super
		super.setEnabled(enabled)

		// Check if the view has been initialized yet or not
		if (!this::parent.isInitialized)
		{
			bufferedEnableStatus = enabled
			return
		}

		// Get the parent layout
		//val parent = checkBox.parent as View

		// Set the alpha of the parent
		parent.alpha = if (enabled) 1.0f else 0.25f
	}

	/**
	 * Setup the color of the check box.
	 */
	private fun setupCheckBoxColor()
	{
		// Create a shared preferences
		val shared = NacSharedPreferences(context)

		// Get the colors for the boolean states
		val colors = intArrayOf(shared.themeColor, Color.LTGRAY)

		// Get the IDs of the two states
		val states = arrayOf(intArrayOf(android.R.attr.state_checked), intArrayOf(-android.R.attr.state_checked))

		// Set the state list
		checkBox.buttonTintList = ColorStateList(states, colors)
	}

	companion object
	{

		/**
		 * Default value.
		 */
		private const val DEFAULT_VALUE = true

	}

}